File: c_clang/agent.c
#include "agent.h"

agent_t *new_agent()
{
    agent_t *agent;

    agent = (agent_t *)malloc(sizeof(agent_t));
    if (agent == NULL)
        exit(EXIT_FAILURE);

    // なんかいい方法ないかね
    agent->init_qvalue = 10000.0;
    agent->actions[0] = -10.0;
    agent->actions[1] = 10.0;

    agent->x_limits[0] = -2.0;
    agent->x_limits[1] = 2.0;
    agent->theta_limits[0] = -M_PI;
    agent->theta_limits[1] = M_PI;
    agent->xdot_limits[0] = -2.0;
    agent->xdot_limits[1] = 2.0;
    agent->thetadot_limits[0] = -10.0;
    agent->thetadot_limits[1] = 10.0;

    agent->x_num = 4;
    agent->theta_num = 90;
    agent->xdot_num = 10;
    agent->thetadot_num = 50;

    agent->x_bins = make_bins(agent->x_limits, agent->x_num);
    agent->theta_bins = make_bins(agent->theta_limits, agent->theta_num);
    agent->xdot_bins = make_bins(agent->xdot_limits, agent->xdot_num);
    agent->thetadot_bins = make_bins(agent->thetadot_limits, agent->thetadot_num);

    agent->alpha = 0.1;
    agent->gamma = 0.999;
    agent->eps = 0.1;
    agent->qtable = make_qtable(
        agent->x_num * agent->theta_num * agent->xdot_num * agent->thetadot_num,
        2,
        agent->init_qvalue);
    return agent;
}

// eps-greedy
double decide_action(agent_t *agent, state_t s)
{
    int s_idx, max_idx;

    if (((double)rand() / (double)RAND_MAX) < agent->eps)
    {
        int idx = (int)(rand() * 2.0 / (1.0 + RAND_MAX));
        return agent->actions[idx];
    }

    s_idx = s_index(agent, s);
    max_idx = argmax(agent->qtable[s_idx], 2);
    return agent->actions[max_idx];
}

void agent_learn(agent_t *agent, state_t s, double a, double r, state_t s_next)
{
    int s_idx = s_index(agent, s);
    int a_idx = find_idx(agent->actions, 2, a);
    int s_next_idx = s_index(agent, s_next);

    agent->qtable[s_idx][a_idx] =
        (1.0 - agent->alpha) * agent->qtable[s_idx][a_idx] +
        agent->alpha * (r + agent->gamma * find_max(agent->qtable[s_next_idx], 2));
    return;
}

void agent_set_test_params(agent_t *agent)
{
    agent->alpha = 0.0;
    agent->eps = 0.0;
    return;
}

double *make_bins(double limits[], int num)
{
    int i;
    double width = (limits[1] - limits[0]) / (double)(num - 2);
    double *bins;

    bins = (double *)malloc(sizeof(double) * (num - 1));
    if (bins == NULL)
        exit(EXIT_FAILURE);

    for (i = 0; i < num - 1; i++)
        bins[i] = limits[0] + width * i;

    return bins;
}

double **make_qtable(int s_num, int a_num, double init_qvalue)
{
    int i, j;
    double **qtable;
    qtable = (double **)malloc(sizeof(double *) * s_num);
    if (qtable == NULL)
        exit(EXIT_FAILURE);

    for (i = 0; i < s_num; i++)
    {
        qtable[i] = (double *)malloc(sizeof(double) * a_num);
        if (qtable[i] == NULL)
            exit(EXIT_FAILURE);

        for (j = 0; j < a_num; j++)
        {
            qtable[i][j] = init_qvalue;
        }
    }
    return qtable;
}

int digitize(double *bins, int bins_len, double x)
{
    int i;
    for (i = 0; i < bins_len; i++)
    {
        if (x < bins[i])
            return i;
    }
    return bins_len;
}

int s_index(agent_t *agent, state_t s)
{
    int x_idx = digitize(agent->x_bins, agent->x_num - 1, s[0]);
    int theta_idx = digitize(agent->theta_bins, agent->theta_num - 1, s[1]);
    int xdot_idx = digitize(agent->xdot_bins, agent->xdot_num - 1, s[2]);
    int thetadot_idx = digitize(agent->thetadot_bins, agent->thetadot_num - 1, s[3]);
    return x_idx + agent->x_num * (theta_idx + agent->theta_num * (xdot_idx + agent->xdot_num * thetadot_idx));
}

int argmax(double doubles[], int len)
{
    int i;
    int idx = 0;
    double max_val = doubles[0];
    for (i = 0; i < len; i++)
    {
        if (doubles[i] > max_val)
        {
            idx = i;
            max_val = doubles[i];
        }
    }
    return idx;
}

int find_idx(double doubles[], int len, double x)
{
    int i;
    for (i = 0; i < len; i++)
    {
        if (x == doubles[i])
        {
            return i;
        }
    }
    // めんどくさいので強制終了
    exit(EXIT_FAILURE);
    return 0;
}

double find_max(double doubles[], int len)
{
    int i;
    double max = doubles[0];
    for (i = 0; i < len; i++)
    {
        if (doubles[i] > max)
        {
            max = doubles[i];
        }
    }
    return max;
}

File: haskell/src/agent.hs
module Agent
    ( Params
    , defaultParams
    , defaultTestParams
    , Qtable
    , newQtable
    , action
    , learn
    )
where

import qualified Data.List                     as L
import qualified Data.Vector                   as V
import qualified Data.Vector.Generic           as G
import qualified Data.Vector.Generic.Mutable   as GM
import qualified Data.Vector.Unboxed           as VU
import qualified Data.Vector.Mutable           as VM
import qualified Data.Vector.Unboxed.Mutable   as VUM
import qualified System.Random                 as R
import qualified Env                           as E


-- Q-value の初期値
initQvalue = 10000.0

-- 行動の候補
actions = [-10.0, 10.0]

-- 状態分割の下限と上限
xLimits = [-2.0, 2.0]
thetaLimits = [-pi, pi]
xdotLimits = [-2.0, 2.0]
thetadotLimits = [-10.0, 10.0]

-- 状態の分割数
xNum = 4
thetaNum = 40
xdotNum = 10
thetadotNum = 50
maxStateLen = xNum * thetaNum * xdotNum * thetadotNum

makeBins :: [Double] -> Int -> [Double]
makeBins lims num =
    [lims !! 0, lims !! 0 + (lims !! 1 - lims !! 0) / fromIntegral (num - 2) .. lims
        !! 1]

-- 状態分割の bins を生成
xBins = makeBins xLimits xNum
thetaBins = makeBins thetaLimits thetaNum
xdotBins = makeBins xdotLimits xdotNum
thetadotBins = makeBins thetadotLimits thetadotNum

-- 学習に使うパラメータ
data Params = Params { alpha :: Double, gamma :: Double, epsilon :: Double }

defaultParams = Params 0.1 0.999 0.1
defaultTestParams = Params 0.0 0.999 0.0


-- 思い切って1次元配列でいいことにした
type Qtable = IO (VUM.IOVector Double)

newQtable :: IO (VUM.IOVector Double)
newQtable = VUM.replicate (maxStateLen * length actions) initQvalue


action :: Params -> Qtable -> E.State -> IO Double
action params qtable s = do
    r <- R.randomIO :: IO Double
    if r < epsilon params
        then do
            aIdx <- R.randomRIO (0, length actions) :: IO Int
            return $ actions !! aIdx
        else do
            let sIdx = getSIdx s
            actions' <- sliceQtable qtable sIdx
            aIdx     <- argmax actions'
            return $ actions !! aIdx

learn :: Params -> Qtable -> E.State -> E.Action -> E.Reward -> E.State -> IO ()
learn params qtable s a r sNext = do
    let sIdx     = getSIdx s
    let aIdx     = getAIdx a
    let sNextIdx = getSIdx sNext
    table  <- qtable
    qValue <- VUM.unsafeRead table $ getIdx sIdx aIdx
    sli    <- sliceQtable qtable sNextIdx
    maxQ   <- max' sli
    VUM.unsafeWrite table (getIdx sIdx aIdx)
        $ (1.0 - alpha params)
        * qValue
        + alpha params
        * (r + gamma params * maxQ)

digitize :: Double -> [Double] -> Int
digitize x bins = length $ takeWhile (< x) bins

getSIdx :: E.State -> Int
getSIdx s =
    xIdx + xNum * (thetaIdx + thetaNum * (xdotIdx + xdotNum * thetadotIdx))
  where
    (x, theta, xdot, thetadot) = E.tupleState s
    xIdx                       = digitize x xBins
    thetaIdx                   = digitize theta thetaBins
    xdotIdx                    = digitize xdot xdotBins
    thetadotIdx                = digitize thetadot thetadotBins

getAIdx :: E.Action -> Int
getAIdx a = let Just x = L.elemIndex a actions in x

-- TODO length actions == 2 を利用しているがちゃんとしたアルゴリズムにしたい
argmax :: VUM.IOVector Double -> IO Int
argmax vec = do
    a <- GM.unsafeRead vec 0
    b <- GM.unsafeRead vec 1
    if a > b then return 0 else return 1


sliceQtable :: Qtable -> Int -> IO (VUM.IOVector Double)
sliceQtable qtable sIdx = VUM.slice sIdx (length actions) <$> qtable

getIdx :: Int -> Int -> Int
getIdx sIdx aIdx = sIdx * length actions + aIdx

-- TODO length actions == 2 を利用しているがちゃんとしたアルゴリズムにしたい
max' :: VUM.IOVector Double -> IO Double
max' vec = do
    a <- GM.unsafeRead vec 0
    b <- GM.unsafeRead vec 1
    if a > b then return a else return b

